// @ubti/diligence-fabric-sdk/build/module/lib/services/Verify/VerifyService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
import { LoggerService } from "../Logger/LoggerService";
export class VerifyService {
    tenantId;
    appId;
    appEnvironmentCODE;
    headers;
    userId;
    constructor(appId, tenantId, appEnvironmentCODE) {
        this.tenantId = tenantId;
        this.appId = appId;
        this.appEnvironmentCODE = appEnvironmentCODE;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId, 0);
        this.tenantId = parseInt(decodedData?.TenantID, 0);
    }
    async IsEmailVerfied(emailAddress) {
        LoggerService.Logger.info("Validate Email Address " + emailAddress);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.VerifyService.IsEmailVerified,
            body: emailAddress,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async ReVerifyEmail(requestBody) {
        LoggerService.Logger.info("ReVerify Email " + requestBody);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.VerifyService.ReVerify,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async Verify(token) {
        LoggerService.Logger.info("Verify Email " + token);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.VerifyService.Verify,
            body: JSON.stringify(token),
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
