// @ubti/diligence-fabric-sdk/build/main/lib/services/Verify/VerifyService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifyService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
const LoggerService_1 = require("../Logger/LoggerService");
class VerifyService {
    constructor(appId, tenantId, appEnvironmentCODE) {
        this.tenantId = tenantId;
        this.appId = appId;
        this.appEnvironmentCODE = appEnvironmentCODE;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    async IsEmailVerfied(emailAddress) {
        LoggerService_1.LoggerService.Logger.info("Validate Email Address " + emailAddress);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.VerifyService.IsEmailVerified,
            body: emailAddress,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async ReVerifyEmail(requestBody) {
        LoggerService_1.LoggerService.Logger.info("ReVerify Email " + requestBody);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.VerifyService.ReVerify,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async Verify(token) {
        LoggerService_1.LoggerService.Logger.info("Verify Email " + token);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.VerifyService.Verify,
            body: JSON.stringify(token),
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
exports.VerifyService = VerifyService;
