// @ubti/diligence-fabric-sdk/build/main/lib/services/Application/AppMenu/AppMenuService.js
"use strict";
/**
 * Service class for managing application menus.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuService = void 0;
const config_1 = require("../../../../config");
const Request_1 = require("../../../../core/Request");
// eslint-disable-next-line functional/no-class
class MenuService {
    constructor(appId, tenantId) {
        this.headers = {};
        this.userId = 0;
        this.tenantId = 0;
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        var _a, _b;
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt((_a = decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId) !== null && _a !== void 0 ? _a : '0', 10);
        this.tenantId = parseInt((_b = decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID) !== null && _b !== void 0 ? _b : '0', 10);
    }
    /**
     * Retrieves the application menu or a specific menu item by ID.
     * @param requestBody The ID of the menu item to retrieve.
     * @returns The application menu or the specific menu item if an ID is provided, or undefined if not found.
     */
    get(requestBody) {
        // Implement the login logic here
        // Set the isLoggedIn property accordingly
        // LoggerService.Logger.info("Initialized getUserAppRole")
        const url = requestBody.AppMenuID ? config_1.SdkConfig.ServiceConfig.AppMenuService.GET : config_1.SdkConfig.ServiceConfig.AppMenuService.GETALL;
        requestBody.AppID = this.appId;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
    update(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.AppMenuService.UPDATE;
        requestBody.AppID = this.appId;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
    /**
     * Deletes an application menu item by ID.
     * @param MenuID The ID of the menu item to delete.
     * @returns True if the menu item was successfully deleted, false if not found.
     */
    delete(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.AppMenuService.DELETE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    /**
     * Creates an application menu item by ID.
     * @param MenuID The ID of the menu item to create.
     * @returns True if the menu item was successfully created, false if not found.
     */
    create(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.AppMenuService.CREATE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
exports.MenuService = MenuService;
