// @ubti/diligence-fabric-sdk/build/main/lib/services/Application/AppRole/AppRoleService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationRoleService = void 0;
const config_1 = require("../../../../config");
const Request_1 = require("../../../../core/Request");
const LoggerService_1 = require("../../Logger/LoggerService");
// eslint-disable-next-line functional/no-class
class ApplicationRoleService {
    constructor(appId, tenantId, appEnvironmentCODE) {
        this.tenantId = tenantId;
        this.appId = appId;
        this.appEnvironmentCODE = appEnvironmentCODE;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    async getUserAppRole(requestBody, mode = 'ALL') {
        const url = config_1.SdkConfig.ServiceConfig.UserAppRoleService.GETALL;
        requestBody.tenantID = this.tenantId;
        requestBody.calledBy = this.userId;
        requestBody.appID = this.appId;
        requestBody.appEnvironmentCODE = this.appEnvironmentCODE;
        const response = await (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
        if (mode !== 'ALL') {
            response.Result = response.Result.filter(x => x.UserID == this.userId);
        }
        return response;
    }
    async getAppRole(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserAppRoleService.GETALL;
        requestBody.tenantID = this.tenantId;
        requestBody.calledBy = this.userId;
        requestBody.appID = this.appId;
        const response = await (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
        return response;
    }
    getAppMenuAppRole(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.AppMenuAppRoleService.GET;
        requestBody.tenantID = this.tenantId;
        requestBody.calledBy = this.userId;
        requestBody.appID = this.appId;
        requestBody.appEnvironmentCODE = this.appEnvironmentCODE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    get(requestBody) {
        const url = requestBody.AppID ? config_1.SdkConfig.ServiceConfig.ApplicationService.GET : config_1.SdkConfig.ServiceConfig.ApplicationService.GETALL;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async getAllAccessibleMenus(useNewFormat = false) {
        var _a;
        let finalAppMenuResponse = [], appMenus = [];
        const requestBody = {
            tenantID: this.tenantId,
            calledBy: this.userId,
            appID: this.appId,
            appEnvironmentCODE: this.appEnvironmentCODE
        };
        const userRoleList = await this.getUserAppRole(requestBody);
        LoggerService_1.LoggerService.Logger.info("User Role List Response " + userRoleList);
        const filteredResults = userRoleList.Result.filter((x) => x.UserID === this.userId);
        try {
            if (filteredResults.length === 0) {
                throw "User doesn't have any Application Roles associated. Please contact Tenant Admin.";
            }
            const userAppRoleData = JSON.parse((_a = filteredResults[0]) === null || _a === void 0 ? void 0 : _a.AppRoles).map(res => res.AppRoleID);
            const appMenuAppRoleRequest = {
                tenantID: this.tenantId,
                calledBy: this.userId,
                appID: this.appId,
                appEnvironmentCODE: this.appEnvironmentCODE
            };
            const appmenuAppRoleResponse = await this.getAppMenuAppRole(appMenuAppRoleRequest);
            const appmenuAppRoleList = appmenuAppRoleResponse.Result;
            for (let i = 0; i < appmenuAppRoleList.length; i++) {
                const data = JSON.parse(appmenuAppRoleList[i].AppRoles);
                const appMenuRole = data.map(x => x.AppRoleID);
                if (appMenuRole.filter((x) => userAppRoleData.includes(x)).length !== 0) {
                    finalAppMenuResponse.push(appmenuAppRoleList[i]);
                }
            }
            if (!useNewFormat) {
                appMenus = finalAppMenuResponse.map(x => ({
                    AppMenuID: x.AppMenuID,
                    AppMenuLabel: x.AppMenuLabel,
                    ParenAppMenuID: x.ParentAppMenuID,
                    AppMenuActionURL: x.AppMenuActionURL,
                    childMenus: []
                }));
                const childMenus = appMenus.filter(res => res.ParenAppMenuID != 0).map(x => ({
                    AppMenuID: x.AppMenuID,
                    AppMenuLabel: x.AppMenuLabel,
                    ParentAppMenuID: x.ParenAppMenuID,
                    AppMenuActionURL: x.AppMenuActionURL,
                    child_Menus: []
                }));
                const child_Menus = childMenus.filter(res => res.ParentAppMenuID != 0).map(x => ({
                    AppMenuID: x.AppMenuID,
                    AppMenuLabel: x.AppMenuLabel,
                    ParentAppMenuID: x.ParentAppMenuID,
                    AppMenuActionURL: x.AppMenuActionURL
                }));
                for (let k = 0; k < appMenus.length; k++) {
                    const tempData = childMenus.filter(x => x.ParentAppMenuID == appMenus[k].AppMenuID);
                    appMenus[k].childMenus = tempData;
                    for (let j = 0; j < appMenus[k].childMenus.length; j++) {
                        const tempData1 = child_Menus.filter(x => x.ParentAppMenuID == appMenus[k].childMenus[j].AppMenuID);
                        appMenus[k].childMenus[j].child_Menus = tempData1;
                    }
                }
            }
            else {
                // New structured format (hierarchical)
                const formattedMenus = finalAppMenuResponse.map(x => ({
                    AppMenuID: x.AppMenuID,
                    AppMenuLabel: x.AppMenuLabel,
                    ParentAppMenuID: x.ParentAppMenuID,
                    AppMenuActionURL: x.AppMenuActionURL,
                    childMenus: []
                }));
                const menuMap = new Map();
                formattedMenus.forEach(menu => menuMap.set(menu.AppMenuID, menu));
                formattedMenus.forEach(menu => {
                    if (menu.ParentAppMenuID && menuMap.has(menu.ParentAppMenuID)) {
                        const parent = menuMap.get(menu.ParentAppMenuID);
                        parent.childMenus.push(menu);
                    }
                });
                appMenus = formattedMenus.filter(menu => !menu.ParentAppMenuID);
            }
            return appMenus;
        }
        catch (error) {
            LoggerService_1.LoggerService.Logger.error(error);
            return error;
        }
    }
    async isMenuAuthorized(appMenuURLPath) {
        const appRequestBody = {
            TenantID: this.tenantId,
            AppID: this.appId
        };
        const checkApp = await this.get(appRequestBody);
        const menuResult = await this.getAllAccessibleMenus();
        if (Array.isArray(menuResult)) {
            if (menuResult.length === 0) {
                return { status: 404, isAuthorized: false, empty: true, message: "No Menus were found! Please contact your Tenant Admin." };
            }
            if (Boolean(menuResult.find(x => x.AppMenuActionURL && new URL(x.AppMenuActionURL).pathname === appMenuURLPath || appMenuURLPath === '/'))) {
                return { status: 200, isAuthorized: true, empty: false, message: "Authorized" };
            }
        }
        else {
            return { status: 404, isAuthorized: false, empty: true, message: "No Menus were found! Please contact your Tenant Admin." };
        }
        return { status: 403, isAuthorized: false, empty: false, message: "Forbidden! You are not allowed to access this page." };
    }
}
exports.ApplicationRoleService = ApplicationRoleService;
