// @ubti/diligence-fabric-sdk/build/main/lib/services/Application/ApplicationService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
// eslint-disable-next-line functional/no-class
class ApplicationService {
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    get(requestBody) {
        const url = requestBody.AppID ? config_1.SdkConfig.ServiceConfig.ApplicationService.GET : config_1.SdkConfig.ServiceConfig.ApplicationService.GETALL;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAll(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.ApplicationService.GETALL;
        requestBody.TenantID = this.tenantId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    create(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.ApplicationService.CREATE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    update(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.ApplicationService.UPDATE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'PUT',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    delete(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.ApplicationService.DELETE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getLogo(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.ApplicationService.GETLOGO;
        requestBody.TenantID = this.tenantId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
exports.ApplicationService = ApplicationService;
