// @ubti/diligence-fabric-sdk/build/module/lib/services/Application/AppMenu/AppMenuService.js
/**
 * Service class for managing application menus.
 */
import { SdkConfig } from "../../../../config";
import { request } from "../../../../core/Request";
// eslint-disable-next-line functional/no-class
export class MenuService {
    headers = {};
    userId = 0;
    tenantId = 0;
    appId;
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId ?? '0', 10);
        this.tenantId = parseInt(decodedData?.TenantID ?? '0', 10);
    }
    /**
     * Retrieves the application menu or a specific menu item by ID.
     * @param requestBody The ID of the menu item to retrieve.
     * @returns The application menu or the specific menu item if an ID is provided, or undefined if not found.
     */
    get(requestBody) {
        // Implement the login logic here
        // Set the isLoggedIn property accordingly
        // LoggerService.Logger.info("Initialized getUserAppRole")
        const url = requestBody.AppMenuID ? SdkConfig.ServiceConfig.AppMenuService.GET : SdkConfig.ServiceConfig.AppMenuService.GETALL;
        requestBody.AppID = this.appId;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
    update(requestBody) {
        const url = SdkConfig.ServiceConfig.AppMenuService.UPDATE;
        requestBody.AppID = this.appId;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
    /**
     * Deletes an application menu item by ID.
     * @param MenuID The ID of the menu item to delete.
     * @returns True if the menu item was successfully deleted, false if not found.
     */
    delete(requestBody) {
        const url = SdkConfig.ServiceConfig.AppMenuService.DELETE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    /**
     * Creates an application menu item by ID.
     * @param MenuID The ID of the menu item to create.
     * @returns True if the menu item was successfully created, false if not found.
     */
    create(requestBody) {
        const url = SdkConfig.ServiceConfig.AppMenuService.CREATE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
