// @ubti/diligence-fabric-sdk/build/main/lib/services/User/UserService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
// const logger = require('./logger');
// eslint-disable-next-line functional/no-class
class UserService {
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    get(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserService.GET;
        requestBody.tenantID = this.tenantId;
        requestBody.calledBy = this.userId;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAll(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserService.GETALL;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    Create(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserService.CREATE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    update(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserService.UPDATE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'PUT',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    delete(requestBody) {
        const url = config_1.SdkConfig.ServiceConfig.UserService.DELETE;
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
exports.UserService = UserService;
