// @ubti/diligence-fabric-sdk/build/module/config.js
const isBrowser = typeof window !== 'undefined';
// let config = {} as any;
if (!isBrowser) {
    import('dotenv/config');
    // import('config').then((value) => {
    //     config = value;
    // })
}
export const getEnvVariable = (key, defaultValue) => {
    if (isBrowser) {
        return defaultValue;
    }
    else {
        // On server side, return any environment variable
        const rtnValue = process.env[key] || defaultValue;
        console.log(`Checking value on Server Side ${key} value is ${rtnValue}`);
        return rtnValue;
    }
};
const API_VERSION = getEnvVariable('API_VERSION', 'v1');
export const SdkConfig = {
    ServiceConfig: {
        BaseURL: 'https://ubti-d-df-api.azurewebsites.net',
        LoginServiceURL: `/api/${API_VERSION}/Auth/DownstreamAppLogin`,
        AuthService: {
            CHANGEPASS: `/api/${API_VERSION}/User/ChangeUserPassword`,
            FORGOTPASS: `/api/${API_VERSION}/Auth/ForgotPassword`,
            GENERATETOKEN: `/api/${API_VERSION}/Auth/generate-token`,
            ValidateToken: `/api/${API_VERSION}/Auth/ValidateToken`,
            ValidateTenant: `/api/${API_VERSION}/User/ValidateTenantDetails`,
            Register: `/api/${API_VERSION}/Auth/Register`,
            IsDomainExist: `/api/${API_VERSION}/Auth/IsDomainExist`,
            IsRegistered: `/api/${API_VERSION}/Auth/IsRegistered`
        },
        VerifyService: {
            IsEmailVerified: `/api/${API_VERSION}/Verify/IsEmailVerified`,
            ReVerify: `/api/${API_VERSION}/Verify/ReVerify`,
            Verify: `/api/${API_VERSION}/Verify`
        },
        AuthenticationTypeService: {
            GET: `/api/${API_VERSION}/AuthenticationType/GetAuthenticationType`,
            GETNEW: `/api/${API_VERSION}/AuthenticationType/GetDownStreamAuthenticationType`
        },
        OrganizationService: {
            GET: `/api/${API_VERSION}/Organization/GetOrganization`,
            GETALL: `/api/${API_VERSION}/Organization/GetAllOrganizationsByTenantID`,
            CREATE: `/api/${API_VERSION}/Organization/Organization`,
            UPDATE: `/api/${API_VERSION}/Organization`,
            DELETE: `/api/${API_VERSION}/Organization/DeleteOrganization`
        },
        ApplicationService: {
            GET: `/api/${API_VERSION}/App/GetApp`,
            GETALL: `/api/${API_VERSION}/App/GetAllApps`,
            CREATE: `/api/${API_VERSION}/App/App`,
            UPDATE: `/api/${API_VERSION}/App`,
            DELETE: `/api/${API_VERSION}/App/DeleteApp`,
            GETLOGO: `/api/${API_VERSION}/App/GetDownstreamAppLogo`,
        },
        AppModuleService: {
            GET: `/api/${API_VERSION}/AppMenu/GetAppMenu`,
            GETALL: `/api/${API_VERSION}/AppMenu/GetAppMenuList`,
            CREATE: `/api/${API_VERSION}/AppMenu/InsertAppMenu`,
            UPDATE: `/api/${API_VERSION}/AppMenu/UpdateAppMenu`,
            DELETE: `/api/${API_VERSION}/AppMenu/DeleteAppMenu`
        },
        AppMenuService: {
            GET: `/api/${API_VERSION}/AppMenu/GetAppMenu`,
            GETALL: `/api/${API_VERSION}/AppMenu/GetAppMenuList`,
            CREATE: `/api/${API_VERSION}/AppMenu/InsertAppMenu`,
            UPDATE: `/api/${API_VERSION}/AppMenu/UpdateAppMenu`,
            DELETE: `/api/${API_VERSION}/AppMenu/DeleteAppMenu`
        },
        UserService: {
            GET: `/api/${API_VERSION}/User/GetUser`,
            VALIDATE: `/api/${API_VERSION}/User/ValidateUser`,
            FORGOTPASS: `/api/${API_VERSION}/User/ChangeUserPassword`,
            VALIDATERESETTOKEN: `/api/${API_VERSION}/User/ValidateResetPasswordToken`,
            GETALL: `/api/${API_VERSION}/User/GetAllUsers`,
            CREATE: `/api/${API_VERSION}/User/User`,
            UPDATE: `/api/${API_VERSION}/User`,
            DELETE: `/api/${API_VERSION}/User/DeleteUser`
        },
        UserAppRoleService: {
            GETALL: `/api/${API_VERSION}/UserAppRole/GetUserAppRoles`,
            CREATE: `/api/${API_VERSION}/UserAppRole/InsertUserAppRole`
        },
        AppMenuAppRoleService: {
            GET: `/api/${API_VERSION}/AppMenuAppRole/GetAppMenuAppRoles`,
            GETALL: "",
            CREATE: "",
            UPDATE: "",
            DELETE: ""
        },
        DFLoggerService: {
            CREATE: `/api/${API_VERSION}/Log/Insert`,
        }
    },
    APIVersion: "",
    WithCredentials: false,
    Headers: undefined,
    Token: ""
};
export function loadConfig() {
    try {
        // eslint-disable-next-line functional/no-let
        let appId = null, tenantId = null;
        // if (config.has('appId')) appId = config.get('appId');
        // if (config.has('tenantId')) tenantId = config.get('tenantId');
        const userConfig = {
            appId: parseInt(getEnvVariable('DF_APP_ID', appId)),
            tenantId: parseInt(getEnvVariable('DF_TENANT_ID', tenantId)),
            apiUrl: getEnvVariable('DF_API_URL', 'https://ubti-d-df-api.azurewebsites.net'),
            apiVersion: getEnvVariable('DF_API_VERSION', 'v1'),
            appEnvironmentCODE: getEnvVariable('DF_AppEnvironmentCODE', 'No value')
        };
        // Assign the userConfig values to SdkConfig
        SdkConfig.ServiceConfig.BaseURL = userConfig.apiUrl;
        SdkConfig.APIVersion = userConfig.apiVersion;
        return userConfig;
    }
    catch (error) {
        console.error('Error loading configuration:', error);
        throw Error("Error loading configuration");
    }
}
export function configure(conf) {
    try {
        // eslint-disable-next-line functional/no-let
        if (!isBrowser && conf == undefined) {
            return loadConfig();
        }
        else {
            const userConfig = {
                appId: parseInt(getEnvVariable('DF_APP_ID', conf.DF_APP_ID.toString())),
                tenantId: parseInt(getEnvVariable('DF_TENANT_ID', conf.DF_TENANT_ID.toString())),
                apiUrl: getEnvVariable('DF_API_URL', conf.DF_API_URL.toString()),
                appEnvironmentCODE: getEnvVariable('DF_AppEnvironmentCODE', conf.DF_AppEnvironmentCODE.toString()),
                apiVersion: getEnvVariable('DF_API_VERSION', conf.DF_API_VERSION.toString()),
            };
            SdkConfig.ServiceConfig.BaseURL = userConfig.apiUrl;
            SdkConfig.APIVersion = userConfig.apiVersion;
            return userConfig;
        }
    }
    catch (error) {
        console.error('Error loading configuration:', error);
        throw Error("Error loading configuration");
    }
}
