// @ubti/diligence-fabric-sdk/build/main/lib/DiligenceFabricClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiligenceFabricClient = void 0;
const ApplicationService_1 = require("./services/Application/ApplicationService");
const AppRoleService_1 = require("./services/Application/AppRole/AppRoleService");
const AppMenuService_1 = require("./services/Application/AppMenu/AppMenuService");
const UserService_1 = require("./services/User/UserService");
const AuthService_1 = require("./services/Auth/AuthService");
const config_1 = require("../config");
const AuthenticationTypeService_1 = require("./services/AuthenticationType/AuthenticationTypeService");
const DFLoggerService_1 = require("./services/DFLogger/DFLoggerService");
const VerifyService_1 = require("./services/Verify/VerifyService");
// eslint-disable-next-line functional/no-class
class DiligenceFabricClient {
    constructor(config) {
        this.userToken = null;
        // Load configs
        this.userConfig = (0, config_1.configure)(config);
        // Initialize the service objects
        this.appId = this.userConfig.appId;
        this.tenantId = this.userConfig.tenantId;
        this.appEnvironmentCODE = this.userConfig.appEnvironmentCODE;
        this.menuService = new AppMenuService_1.MenuService(this.appId, this.tenantId);
        this.authService = new AuthService_1.AuthService(this.appId, this.tenantId, this.appEnvironmentCODE);
        this.applicationService = new ApplicationService_1.ApplicationService(this.appId, this.tenantId);
        this.userService = new UserService_1.UserService(this.appId, this.tenantId);
        this.applicationRoleService = new AppRoleService_1.ApplicationRoleService(this.appId, this.tenantId, this.appEnvironmentCODE);
        this.authenticationTypeService = new AuthenticationTypeService_1.AuthenticationTypeService(this.appId, this.tenantId);
        this.dfLoggerService = new DFLoggerService_1.DFLoggerService(this.appId, this.tenantId);
        this.verifyService = new VerifyService_1.VerifyService(this.appId, this.tenantId, this.appEnvironmentCODE);
    }
    setAuthUser(authUser) {
        this.userToken = authUser.Token;
    }
    getUserToken() {
        if (!this.userToken) {
            throw new Error('User token not set. Please call setAuthUser() first.');
        }
        return this.userToken;
    }
    getDFLoggerService() {
        try {
            this.dfLoggerService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.dfLoggerService;
    }
    getAuthenticationTypeService() {
        return this.authenticationTypeService;
    }
    getApplicationService() {
        this.applicationService.setUserToken(this.getUserToken());
        return this.applicationService;
    }
    getUserService() {
        this.userService.setUserToken(this.getUserToken());
        return this.userService;
    }
    getMenuService() {
        this.menuService.setUserToken(this.getUserToken());
        return this.menuService;
    }
    getApplicationRoleService() {
        this.applicationRoleService.setUserToken(this.getUserToken());
        return this.applicationRoleService;
    }
    getAuthService() {
        try {
            this.authService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.authService;
    }
    getVerifyService() {
        try {
            this.verifyService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.verifyService;
    }
}
exports.DiligenceFabricClient = DiligenceFabricClient;
