// @ubti/diligence-fabric-sdk/build/module/lib/services/User/UserService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
// const logger = require('./logger');
// eslint-disable-next-line functional/no-class
export class UserService {
    headers;
    userId;
    tenantId;
    appId;
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId, 0);
        this.tenantId = parseInt(decodedData?.TenantID, 0);
    }
    get(requestBody) {
        const url = SdkConfig.ServiceConfig.UserService.GET;
        requestBody.tenantID = this.tenantId;
        requestBody.calledBy = this.userId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAll(requestBody) {
        const url = SdkConfig.ServiceConfig.UserService.GETALL;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    Create(requestBody) {
        const url = SdkConfig.ServiceConfig.UserService.CREATE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    update(requestBody) {
        const url = SdkConfig.ServiceConfig.UserService.UPDATE;
        return request(SdkConfig, {
            method: 'PUT',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    delete(requestBody) {
        const url = SdkConfig.ServiceConfig.UserService.DELETE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
