// @ubti/diligence-fabric-sdk/build/module/lib/services/Application/ApplicationService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
// eslint-disable-next-line functional/no-class
export class ApplicationService {
    headers;
    userId;
    tenantId;
    appId;
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId, 0);
        this.tenantId = parseInt(decodedData?.TenantID, 0);
    }
    get(requestBody) {
        const url = requestBody.AppID ? SdkConfig.ServiceConfig.ApplicationService.GET : SdkConfig.ServiceConfig.ApplicationService.GETALL;
        requestBody.TenantID = this.tenantId;
        requestBody.CalledBy = this.userId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAll(requestBody) {
        const url = SdkConfig.ServiceConfig.ApplicationService.GETALL;
        requestBody.TenantID = this.tenantId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    create(requestBody) {
        const url = SdkConfig.ServiceConfig.ApplicationService.CREATE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    update(requestBody) {
        const url = SdkConfig.ServiceConfig.ApplicationService.UPDATE;
        return request(SdkConfig, {
            method: 'PUT',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    delete(requestBody) {
        const url = SdkConfig.ServiceConfig.ApplicationService.DELETE;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getLogo(requestBody) {
        const url = SdkConfig.ServiceConfig.ApplicationService.GETLOGO;
        requestBody.TenantID = this.tenantId;
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
