// @ubti/diligence-fabric-sdk/build/module/lib/services/DFLogger/DFLoggerService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
export class DFLoggerService {
    headers = {};
    userId = 0;
    tenantId = 0;
    appId;
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId, 0);
        this.tenantId = parseInt(decodedData?.TenantID, 0);
    }
    /**
     * creates log in df.
     * @param requestBody The log request parameters.
     * @returns The log inserted success message or returns bad request
     */
    log(logLevel, callingFunction, dynmcFieldsJSON) {
        const url = SdkConfig.ServiceConfig.DFLoggerService.CREATE;
        const requestBody = {
            logLevel: logLevel,
            logTypeCODE: 'MSGLOG',
            callingFunction: callingFunction,
            dynmcFieldsJSON: dynmcFieldsJSON,
            AppID: this.appId,
            TenantID: this.tenantId
        };
        return request(SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
}
