// @ubti/diligence-fabric-sdk/build/module/lib/services/AuthenticationType/AuthenticationTypeService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
import { LoggerService } from "../Logger/LoggerService";
export class AuthenticationTypeService {
    tenantId;
    appId;
    headers;
    userId;
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    getAuthenticationType(requestBody) {
        LoggerService.Logger.info("Returning Authentication Type Configurations");
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthenticationTypeService.GET,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAuthenticationDownStreamType(requestBody) {
        LoggerService.Logger.info("Returning Authentication DownStream Type Configurations");
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthenticationTypeService.GETNEW,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
