// @ubti/diligence-fabric-sdk/build/module/lib/DiligenceFabricClient.js
import { ApplicationService } from './services/Application/ApplicationService';
import { ApplicationRoleService } from './services/Application/AppRole/AppRoleService';
import { MenuService } from './services/Application/AppMenu/AppMenuService';
import { UserService } from './services/User/UserService';
import { AuthService } from './services/Auth/AuthService';
import { configure } from '../config';
import { AuthenticationTypeService } from './services/AuthenticationType/AuthenticationTypeService';
import { DFLoggerService } from './services/DFLogger/DFLoggerService';
import { VerifyService } from './services/Verify/VerifyService';
// eslint-disable-next-line functional/no-class
export class DiligenceFabricClient {
    menuService;
    authService;
    applicationService;
    authenticationTypeService;
    userService;
    applicationRoleService;
    userToken = null;
    appId;
    tenantId;
    appEnvironmentCODE;
    userConfig;
    dfLoggerService;
    verifyService;
    constructor(config) {
        // Load configs
        this.userConfig = configure(config);
        // Initialize the service objects
        this.appId = this.userConfig.appId;
        this.tenantId = this.userConfig.tenantId;
        this.appEnvironmentCODE = this.userConfig.appEnvironmentCODE;
        this.menuService = new MenuService(this.appId, this.tenantId);
        this.authService = new AuthService(this.appId, this.tenantId, this.appEnvironmentCODE);
        this.applicationService = new ApplicationService(this.appId, this.tenantId);
        this.userService = new UserService(this.appId, this.tenantId);
        this.applicationRoleService = new ApplicationRoleService(this.appId, this.tenantId, this.appEnvironmentCODE);
        this.authenticationTypeService = new AuthenticationTypeService(this.appId, this.tenantId);
        this.dfLoggerService = new DFLoggerService(this.appId, this.tenantId);
        this.verifyService = new VerifyService(this.appId, this.tenantId, this.appEnvironmentCODE);
    }
    setAuthUser(authUser) {
        this.userToken = authUser.Token;
    }
    getUserToken() {
        if (!this.userToken) {
            throw new Error('User token not set. Please call setAuthUser() first.');
        }
        return this.userToken;
    }
    getDFLoggerService() {
        try {
            this.dfLoggerService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.dfLoggerService;
    }
    getAuthenticationTypeService() {
        return this.authenticationTypeService;
    }
    getApplicationService() {
        this.applicationService.setUserToken(this.getUserToken());
        return this.applicationService;
    }
    getUserService() {
        this.userService.setUserToken(this.getUserToken());
        return this.userService;
    }
    getMenuService() {
        this.menuService.setUserToken(this.getUserToken());
        return this.menuService;
    }
    getApplicationRoleService() {
        this.applicationRoleService.setUserToken(this.getUserToken());
        return this.applicationRoleService;
    }
    getAuthService() {
        try {
            this.authService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.authService;
    }
    getVerifyService() {
        try {
            this.verifyService.setUserToken(this.getUserToken());
        }
        catch (ex) {
            console.log(ex);
        }
        return this.verifyService;
    }
}
