// @ubti/diligence-fabric-sdk/build/main/lib/services/AuthenticationType/AuthenticationTypeService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticationTypeService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
const LoggerService_1 = require("../Logger/LoggerService");
class AuthenticationTypeService {
    constructor(appId, tenantId) {
        this.tenantId = tenantId;
        this.appId = appId;
    }
    getAuthenticationType(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Returning Authentication Type Configurations");
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthenticationTypeService.GET,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    getAuthenticationDownStreamType(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Returning Authentication DownStream Type Configurations");
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthenticationTypeService.GETNEW,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
}
exports.AuthenticationTypeService = AuthenticationTypeService;
