// @ubti/diligence-fabric-sdk/build/main/lib/services/DFLogger/DFLoggerService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DFLoggerService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
class DFLoggerService {
    constructor(appId, tenantId) {
        this.headers = {};
        this.userId = 0;
        this.tenantId = 0;
        this.tenantId = tenantId;
        this.appId = appId;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    /**
     * creates log in df.
     * @param requestBody The log request parameters.
     * @returns The log inserted success message or returns bad request
     */
    log(logLevel, callingFunction, dynmcFieldsJSON) {
        const url = config_1.SdkConfig.ServiceConfig.DFLoggerService.CREATE;
        const requestBody = {
            logLevel: logLevel,
            logTypeCODE: 'MSGLOG',
            callingFunction: callingFunction,
            dynmcFieldsJSON: dynmcFieldsJSON,
            AppID: this.appId,
            TenantID: this.tenantId
        };
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: url,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers,
        });
    }
}
exports.DFLoggerService = DFLoggerService;
