// @ubti/diligence-fabric-sdk/build/main/lib/services/Auth/AuthService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthService = void 0;
const config_1 = require("../../../config");
const Request_1 = require("../../../core/Request");
const LoggerService_1 = require("../Logger/LoggerService");
// eslint-disable-next-line functional/no-class
class AuthService {
    constructor(appId, tenantId, appEnvironmentCODE) {
        this.tenantId = tenantId;
        this.appId = appId;
        this.appEnvironmentCODE = appEnvironmentCODE;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.userId, 0);
        this.tenantId = parseInt(decodedData === null || decodedData === void 0 ? void 0 : decodedData.TenantID, 0);
    }
    async Register(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Register flow " + requestBody);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthService.Register,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    ValidateToken(token) {
        LoggerService_1.LoggerService.Logger.info("Validate Token " + token);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'GET',
            url: `${config_1.SdkConfig.ServiceConfig.AuthService.ValidateToken}?token=${encodeURIComponent(token)}`,
            headers: this.headers
        });
    }
    ValidateTenant(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Validate Tenant " + requestBody);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthService.ValidateTenant,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    IsDomainExists(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Domain or Registration Exist" + requestBody);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthService.IsDomainExist,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    IsRegistered(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Domain or Registration Exist" + requestBody);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthService.IsRegistered,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async login(requestBody) {
        LoggerService_1.LoggerService.Logger.info("Login " + requestBody.username);
        const authRequestBody = Object.assign(Object.assign({}, requestBody), { tenantID: this.tenantId, appID: this.appId, appEnvironmentCODE: this.appEnvironmentCODE });
        const response = await (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.LoginServiceURL,
            body: authRequestBody,
            mediaType: 'application/json',
        });
        const loginAuthenticationResponse = { Message: response.Message, StatusCode: response.StatusCode, Result: response.Result.find(x => x.TenantID == this.tenantId) };
        return loginAuthenticationResponse;
    }
    forgotPassword(requestBody) {
        LoggerService_1.LoggerService.Logger.info("forgot Password " + requestBody.email);
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'POST',
            url: config_1.SdkConfig.ServiceConfig.AuthService.FORGOTPASS,
            body: requestBody,
            mediaType: 'application/json',
        });
    }
    changePassword(requestBody) {
        return (0, Request_1.request)(config_1.SdkConfig, {
            method: 'PUT',
            url: config_1.SdkConfig.ServiceConfig.AuthService.CHANGEPASS,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    logout() {
        // Implement the logout logic here
        // Reset the isLoggedIn property accordingly
        return;
    }
}
exports.AuthService = AuthService;
