// @ubti/diligence-fabric-sdk/build/module/lib/services/Auth/AuthService.js
import { SdkConfig } from "../../../config";
import { request } from "../../../core/Request";
import { LoggerService } from "../Logger/LoggerService";
// eslint-disable-next-line functional/no-class
export class AuthService {
    tenantId;
    appId;
    appEnvironmentCODE;
    headers;
    userId;
    constructor(appId, tenantId, appEnvironmentCODE) {
        this.tenantId = tenantId;
        this.appId = appId;
        this.appEnvironmentCODE = appEnvironmentCODE;
    }
    setUserToken(token) {
        this.headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` };
        const decodedData = JSON.parse(atob(token.split('.')[1]));
        this.userId = parseInt(decodedData?.userId, 0);
        this.tenantId = parseInt(decodedData?.TenantID, 0);
    }
    async Register(requestBody) {
        LoggerService.Logger.info("Register flow " + requestBody);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthService.Register,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    ValidateToken(token) {
        LoggerService.Logger.info("Validate Token " + token);
        return request(SdkConfig, {
            method: 'GET',
            url: `${SdkConfig.ServiceConfig.AuthService.ValidateToken}?token=${encodeURIComponent(token)}`,
            headers: this.headers
        });
    }
    ValidateTenant(requestBody) {
        LoggerService.Logger.info("Validate Tenant " + requestBody);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthService.ValidateTenant,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    IsDomainExists(requestBody) {
        LoggerService.Logger.info("Domain or Registration Exist" + requestBody);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthService.IsDomainExist,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    IsRegistered(requestBody) {
        LoggerService.Logger.info("Domain or Registration Exist" + requestBody);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthService.IsRegistered,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    async login(requestBody) {
        LoggerService.Logger.info("Login " + requestBody.username);
        const authRequestBody = { ...requestBody, tenantID: this.tenantId, appID: this.appId, appEnvironmentCODE: this.appEnvironmentCODE };
        const response = await request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.LoginServiceURL,
            body: authRequestBody,
            mediaType: 'application/json',
        });
        const loginAuthenticationResponse = { Message: response.Message, StatusCode: response.StatusCode, Result: response.Result.find(x => x.TenantID == this.tenantId) };
        return loginAuthenticationResponse;
    }
    forgotPassword(requestBody) {
        LoggerService.Logger.info("forgot Password " + requestBody.email);
        return request(SdkConfig, {
            method: 'POST',
            url: SdkConfig.ServiceConfig.AuthService.FORGOTPASS,
            body: requestBody,
            mediaType: 'application/json',
        });
    }
    changePassword(requestBody) {
        return request(SdkConfig, {
            method: 'PUT',
            url: SdkConfig.ServiceConfig.AuthService.CHANGEPASS,
            body: requestBody,
            mediaType: 'application/json',
            headers: this.headers
        });
    }
    logout() {
        // Implement the logout logic here
        // Reset the isLoggedIn property accordingly
        return;
    }
}
